File: frontend/src/api/notifications.ts
import type { Notification, NotificationPreference, NotificationPreferenceUpdate } from "../types/notification";
import { client } from "./client";

export const notificationsApi = {
  getAll: () =>
    client.get<Notification[]>("/notifications").then((r) => r.data),

  getUnreadCount: () =>
    client.get<{ count: number }>("/notifications/unread-count").then((r) => r.data),

  markRead: (id: string) =>
    client.patch<Notification>(`/notifications/${id}/read`).then((r) => r.data),

  markAllRead: () =>
    client.patch("/notifications/read-all").then((r) => r.data),

  getPreferences: () =>
    client.get<NotificationPreference>("/users/me/preferences").then((r) => r.data),

  updatePreferences: (data: NotificationPreferenceUpdate) =>
    client.put<NotificationPreference>("/users/me/preferences", data).then((r) => r.data),
};

File: backend/app/api/v1/notifications.py
from __future__ import annotations

from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_current_user
from app.core.database import get_db
from app.models.user import User
from app.schemas.notification import NotificationResponse
from app.services.notification import notification_service

router = APIRouter(prefix="/notifications", tags=["notifications"])


@router.get("", response_model=list[NotificationResponse])
async def list_notifications(
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> list[NotificationResponse]:
    notifications = await notification_service.get_all(db, current_user)
    return [NotificationResponse.model_validate(n) for n in notifications]


@router.get("/unread-count")
async def unread_count(
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> dict:
    count = await notification_service.get_unread_count(db, current_user)
    return {"count": count}


@router.patch("/{notification_id}/read", response_model=NotificationResponse)
async def mark_read(
    notification_id: str,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> NotificationResponse:
    notification = await notification_service.mark_read(db, current_user, notification_id)
    return NotificationResponse.model_validate(notification)


@router.patch("/read-all")
async def mark_all_read(
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> dict:
    await notification_service.mark_all_read(db, current_user)
    return {"message": "All marked read"}
